File: C#/Cypher/Cypher.cs
/*********************************************************************
 * Cypher Program (C#)
 * ------------------------------------------------------------------
 * User enters data into the application and it converts it over to
 * ASCII numbers. This data is saved to a text file, and the user is
 * able to receive data from the text file.
 * 
 * The data is read from the file, and converted back into text to
 * display to the user.
 ********************************************************************/
using System;
using System.IO;

namespace Cypher
{
    class Cypher
    {
        static void Main(string[] args)
        {
            //declare variables
            int option;

            //main program
            Console.WriteLine("Cypher Program");
            Console.WriteLine("==============");

            //call the function for user selection
            option = OptionSelect();

            //switch through returned choice for variable 'option'
            switch (option)
            {
                case '1': WriteFile(); break;
                case '2': ReadFile(); break;
                case '3': break;
                default: Console.WriteLine("\n\n**Error in detecting selection." +
                    " Please restart program and try again."); break;
            }

            //end program sequence
            Console.WriteLine();
            Console.WriteLine("Thank you for using this program. " +
                "Please press Enter to exit.");
            Console.ReadLine();
        }

        //declare functions
        static char OptionSelect()
        {
            //declare variables
            char option = 'a';

            //do loop until user selects the appropriate choice, or '3' to exit
            do
            {
                //display options to user
                Console.WriteLine("1) Write to file");
                Console.WriteLine("2) Read from file");
                Console.WriteLine("3) Exit Program");
                Console.Write("Please choose an option: ");

                try
                {
                    //assign user input to variable option
                    option = Convert.ToChar(Console.ReadLine());

                    //switch through choice to determine option
                    switch (option)
                    {
                        case '1': Console.WriteLine("Write..."); break;
                        case '2': Console.WriteLine("Read..."); break;
                        case '3': break;
                        default: Console.WriteLine("\n\n**Invalid input. Please try again."); break;
                    }
                }
                catch
                {
                    //catch for empty/null entry
                    Console.WriteLine("\n\n**Input must be a number. Please try again.");
                }
            } while (option != '1' && option != '2' && option != '3');

            return option;
        }

        static void WriteFile()
        {
            //declare variables
            string fileName;
            string userText;
            int convertLtr;

            try
            {
                //ask user for file name to be created and take input for the file's text
                Console.WriteLine("Please enter file name to be created:");
                fileName = Console.ReadLine();

                //check if file exists
                if (!File.Exists(fileName + ".txt"))
                {
                    //ask user for input
                    Console.WriteLine("Please enter text:");
                    userText = Console.ReadLine();

                    using (StreamWriter sw = File.CreateText(fileName + ".txt"))
                    {
                        //convert data to ASCII table integers
                        for (int i = 0; i < userText.Length; i++)
                        {
                            convertLtr = userText[i];
                            sw.Write(convertLtr + " ");
                        }
                        sw.Write("\n");
                    }

                    Console.WriteLine($"File '{fileName}.txt' has been successfully created and saved.");
                }
                else
                {
                    //ask user for input
                    Console.WriteLine("Please enter text:");
                    userText = Console.ReadLine();

                    using (StreamWriter sw = File.AppendText(fileName + ".txt"))
                    {
                        //convert data to ASCII table integers
                        for (int i = 0; i < userText.Length; i++)
                        {
                            convertLtr = userText[i];
                            sw.Write(convertLtr + " ");
                        }
                        sw.Write("\n");
                    }
                }
            }catch
            {
                Console.WriteLine("\n\n**Error creating file. Please try again.");
            }

        }

        static void ReadFile()
        {
            //declare variables
            string fileName;
            string textNum = "";
            int textWord;

            try
            {
                //ask user for file name
                Console.WriteLine("Please enter the file name to be read:");
                fileName = Console.ReadLine();

                //check if file exists
                if (!File.Exists(fileName + ".txt"))
                {
                    Console.WriteLine($"\n\n**Error 404: File '{fileName}.txt' not found.");
                }
                else
                {
                    //open file
                    using (StreamReader sr = File.OpenText(fileName + ".txt"))
                    {
                        string s;

                        //loop through each line to collect text
                        while ((s = sr.ReadLine()) != null)
                        {
                            //step through each index on the line of text
                            for (int i = 0; i < s.Length; i++)
                            {
                                //check if text is not a space
                                if(s[i] != ' ')
                                {
                                    if (char.IsNumber(s[i]))
                                    {
                                        textNum += s[i];
                                    }
                                    else
                                    {
                                        Console.WriteLine($"\n\n**Error reading file '{fileName}.txt'. It may be damaged or corrupted.");
                                        return;
                                    }
                                }
                                
                                //check if text is a space
                                if(s[i] == ' ')
                                {
                                    textWord = Convert.ToInt32(textNum);
                                    Console.Write(Convert.ToChar(textWord));
                                    textNum = "";
                                }
                            }
                            Console.WriteLine();
                        }
                        Console.WriteLine("\n\nEnd of file.");
                    }
                }
            }
            catch
            {
                Console.WriteLine("\n\n**Error reading file. It may be damaged or corrupted.");
            }
        }
    }
}

File: C++/Cypher/Cypher.cpp
/**********************************************************************
* Cypher Program
* ---------------------------------------------------------------------
* User enters data into the application and it converts it over to
* ASCII numbers. This data is saved to a text file, and the user is
* able to receive data from the text file.
* 
* The data is read from the file, and converted back into text to
* display to the user.
**********************************************************************/
#include <iostream>
#include <fstream>
#include <string>

//declare functions
char optionSelect() {
    //declare variables
    char option;

    //do loop until user selects appropriate choice
    do {
        std::cout << "1) Write to file" << std::endl;
        std::cout << "2) Read from file" << std::endl;
        std::cout << "3) Exit Program" << std::endl;
        std::cout << std::endl << "Choose an option: ";
        std::cin >> option;

        //switch through choice to determine option
        switch (option) {
        case '1':std::cout << "Write..." << std::endl; break;
        case '2': std::cout << "Read..." << std::endl; break;
        case '3':break;
        default:std::cout << "Invalid input. Please try again." << std::endl; break;
        }
    } while (option != '1' && option != '2' && option != '3');

    return option;
}

void writeFile() {
    //declare variables
    std::fstream file;
    std::string userLine;
    std::string fileName;

    //ask user for file name to be created
    std::cout << "Please enter a file name to be created: " << std::endl;
    std::cin >> fileName;
    fileName += ".txt";

    //open file in read mode
    //file.open("textfile.txt", std::ios::app);
    file.open(fileName, std::ios::app);

    //ask user to enter data
    std::cout << "Please enter text:" << std::endl;
    std::cin.ignore(10000, '\n');
    std::getline(std::cin, userLine);

    //write data to file
    if (file.is_open()) {
        //convert data to ASCII table integers
        for (int i = 0; i < userLine.size(); i++) {
            int convertLtr = userLine[i];
            file << convertLtr << " ";
        }
        file << "\n";
    }

    //close file
    file.close();

    //inform user file has been successfully saved.
    std::cout << fileName << " has been successfully created and saved." << std::endl;
}

void readFile() {
    //declare variables
    std::fstream file;
    std::string textLine;
    std::string textLine2;
    std::string fileName;
    int textWord;

    //ask user for file name to be created
    std::cout << "Please enter a file name to be read: " << std::endl;
    std::cin >> fileName;
    fileName += ".txt";

    //open file in read mode
    file.open(fileName, std::ios::in);

    //read lines from text file and convert to characters.
    if (!file.is_open()) {
        std::cout << "\n\n**Error 404: File '" << fileName << "' not found." << std::endl;
    }
    if (file.is_open()) {
        while (getline(file, textLine, '\n')) {
            for (int x = 0; x < textLine.size(); x++) {
                if (textLine[x] != ' ') {
                    if (isdigit(textLine[x])) {
                        textLine2 += textLine[x];
                    }
                    else {
                        std::cout << "\n\n**Error reading file. It may be damaged or corrupted." << std::endl;
                        return;
                    }
                }
                if (textLine[x] == ' ') {
                    textWord = std::stoi(textLine2);
                    std::cout << (char)textWord;
                    textLine2 = "";
                }
            };
            std::cout << std::endl;
        };
        std::cout << "\n\nEnd of file." << std::endl;
    };

    //close file
    file.close();

    std::cin.get();
};

int main()
{
    //declare variables
    char option;

    //main program
    std::cout << "Cypher Program" << std::endl;
    std::cout << "==============" << std::endl;

    //call function for user selection
    option = optionSelect();

    //switch through returned choice in variable 'option'
    switch (option) {
    case '1':writeFile(); break;
    case '2':readFile(); break;
    case '3':break;
    default: std::cout << "\n\n**Error in detecting selection. Please restart program and try again." << std::endl;; break;
    }

    //end program sequence
    std::cout << std::endl << "Thank you for using this program. Press Enter to exit." << std::endl;
    std::cin.get();
}
